### PR TITLE
Notify Seller when bid was placed

### PR DESCRIPTION
**Pull request by vikas-cis**

### CODE REVIEW

**Comment by AllienWorks** [comment]
Tested, works nicely :+1: 
Needs code review though @kewde

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 1]
```diff
@@ -54,13 +54,17 @@ export class OrderStatusNotifierService implements OnDestroy {
...
   checkOrders(newOrders: any) {
-    this.oldOrders.forEach(order => {
-      newOrders.forEach(newOrder => {
-        if (this.compareOrder(order, newOrder)) {
-          this.notifyNewStatus(newOrder);
-        }
+    if (this.oldOrders.length !== newOrders.length) {
+      this.notifyNewStatus(newOrders.reverse()[0]);
+    } else {
+      this.oldOrders.forEach(order => {
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: This is a bit strange though, why can't you just do newOrders.forEach and then oldOrders.find() using the compare function?

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 2]
```diff
@@ -54,13 +54,17 @@ export class OrderStatusNotifierService implements OnDestroy {
   }
 
   checkOrders(newOrders: any) {
-    this.oldOrders.forEach(order => {
-      newOrders.forEach(newOrder => {
-        if (this.compareOrder(order, newOrder)) {
-          this.notifyNewStatus(newOrder);
-        }
+    if (this.oldOrders.length !== newOrders.length) {
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: This only notifies of the first order that has changed, if two orders changed it won't do anything..

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 3]
```diff
@@ -35,56 +36,64 @@ export class OrderStatusNotifierService implements OnDestroy {
...
-
-    if (!this.oldOrders || this.oldOrders.length === 0) {
-      this.oldOrders = orders.bid.orders;
-    } else {
-      this.checkOrders(orders.bid.orders);
-      this.oldOrders = orders.bid.orders;
+    if (this.oldOrders && this.oldOrders.length) {
+      this.checkOrders(orders.bid.orders.reverse());
     }
+    this.oldOrders = orders.bid.orders.reverse();
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: Is this intended reverse behavior? 
https://www.w3schools.com/jsref/jsref_reverse.asp

reverse() does not return a new array with the reversed order.
It reverses the array you are calling it on.

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 4]
```diff
@@ -35,56 +36,64 @@ export class OrderStatusNotifierService implements OnDestroy {
...
-          this.notifyNewStatus(newOrder);
-        }
+    let compOrders = [];
+    // For New Orders
+    if (this.oldOrders.length !== newOrders.length) {
+      compOrders = this.checkForNewOrders(newOrders);
+    }
+    // OldOrders that are changed
+    this.oldOrders.forEach(oldOrder => {
+      compOrders.push(newOrders.find(newOrder => this.compareOrder(oldOrder, newOrder)))
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: Let me think about this one for a while.

I would approach it in a different way. Take the newOrders, then remove all the oldOrders by your comparison function. 

compOrders = newOrders - oldOrders (using compareOrder).

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 5]
```diff
@@ -35,56 +36,64 @@ export class OrderStatusNotifierService implements OnDestroy {
...
+
+    _.without(compOrders, undefined)
+      .forEach(newOrder => {
+        this.notifyNewStatus(newOrder);
       });
-    });
   }
 
-  compareOrder(order: any, newOrder: any) {
+  private compareOrder(order: any, newOrder: any): any {
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: The name 'compareOrder' is wrong here, you returning true if they are _different_.
compareOrder would by convention return true if they are the same.

A better name is: 'isNewOrder'.

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts (hunk 3 above)
Comment (vikas-cis) [reply src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: hmm... yeah need to reverse first and then assign :)

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 6]
```diff
@@ -19,78 +24,101 @@ export class OrderStatusNotifierService implements OnDestroy {
             'Order rejected'
           ];
   public oldOrders: Array<any>;
-
+  orders: Bid;
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: This is supposed to be an array of Bids.
The Bid model is really mangled, it is supposed to be _one_ bid and not a collection of _bids_.

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 7]
```diff
@@ -19,78 +24,101 @@ export class OrderStatusNotifierService implements OnDestroy {
...
   ) {
     this.log.d('order status notifier service running!');
+    this.loadOrders();
+  }
+
+  loadOrders() {
+    // @TODO need to replace with marketplace command so this should probably gone :)
+    this._marketState.observe('bid')
+      .takeWhile(() => !this.destroyed)
+      .map(o => new Bid(o, this.profile.address))
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: Same here it it supposed to map to an array of bids.

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 8]
```diff
@@ -19,78 +24,101 @@ export class OrderStatusNotifierService implements OnDestroy {
...
 
-  compareOrder(order: any, newOrder: any) {
-    return order.id === newOrder.id &&
-            order.messages.action_button !== newOrder.messages.action_button &&
-            this.actionStatus.includes(newOrder.messages.action_button)
+  private hasOrderChanged(msg: string): boolean {
+    return this.actionStatus.includes(msg)
+  }
+
+  public checkForNewOrders(newOrders: any): any {
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: Should be like:

```
public getOrdersToNotifyFor(newOrders: Bid[]): Bid[] {

File: src/app/core/market/order-status-notifier/order-status-notifier.service.ts [hunk 9]
```diff
@@ -1,96 +1,126 @@
...
+    if (orders.length === 0) {
       return;
     }
-
-    if (!this.oldOrders || this.oldOrders.length === 0) {
-      this.oldOrders = orders.bid.orders;
-    } else {
-      this.checkOrders(orders.bid.orders);
-      this.oldOrders = orders.bid.orders;
+    orders.reverse();
```
Comment (kewde) [inline src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: Move all the reverse()'ing into the bid collection model

Comment (vikas-cis) [reply src/app/core/market/order-status-notifier/order-status-notifier.service.ts]: yup :)

**Pull request merged**